src/transforms/tag_cardinality_limit.rs

```diff
@@ -0,0 +1,459 @@
+        }
+    }
+
+    /// Takes in key and a value corresponding to a tag on an incoming Metric Event.
+    /// If that value is already part of set of accepted values for that key, then simply returns
+    /// true.  If that value is not yet part of the accepted values for that key, checks whether
+    /// we have hit the value_limit for that key yet and if not adds the value to the set of
+    /// accepted values for the key and returns true, otherwise returns false.  A false return
+    /// value indicates to the caller that the value is not accepted for this key, and the
+    /// configured limit_exceeded_action should be taken.
+    fn try_accept_tag(&mut self, key: &String, value: &String) -> bool {
+        if !self.accepted_tags.contains_key(key) {
+            self.accepted_tags.insert(
+                key.clone(),
+                TagValueSet::new(self.config.value_limit, &self.config.mode),
+            );
+        }
+        let tag_value_set = self.accepted_tags.get_mut(key).unwrap();
+
+        if tag_value_set.contains(value) {
+            // Tag value has already been accepted, nothing more to do.
+            return true;
+        }
+
+        // Tag value not yet part of the accepted set.
+        if tag_value_set.len() < self.config.value_limit as usize {
+            // accept the new value
+            tag_value_set.insert(value);
+
+            if tag_value_set.len() == self.config.value_limit as usize {
+                warn!(
+                    "value_limit reached for key {}. New values for this key will be rejected",
+                    key
+                );
+            }
+
+            true
+        } else {
+            // New tag value is rejected.
+            false
+        }
+    }
+}
+
+impl Transform for TagCardinalityLimit {
+    fn transform(&mut self, mut event: Event) -> Option<Event> {
+        match event.as_mut_metric().tags {
+            Some(ref mut tags_map) => {
+                match self.config.limit_exceeded_action {
+                    LimitExceededAction::DropEvent => {
+                        for (key, value) in tags_map {
+                            if !self.try_accept_tag(key, value) {
+                                info!(
+                                    "Rejecting Metric Event containing tag with key: {}, value: {}",
+                                    key, value
+                                );
+                                return None;
+                            }
+                        }
+                    }
+                    LimitExceededAction::DropTag => {
+                        let tags_map_clone = tags_map.clone();
+                        let mut to_delete = Vec::new();
+                        for (key, value) in &tags_map_clone {
+                            if !self.try_accept_tag(key, value) {
+                                info!("Rejecting tag with key: {}, value: {}", key, value);
+                                to_delete.push(key);
```

Comment:
If we clone `key` here, we should be able to avoid cloning the whole tag map upfront.

Reply:
I had tried that before but was still getting errors from the borrow checker.  Looks like explicitly calling .iter() on tags_map in the for-loop description fixed it though.

I guess the issue was that without the explicit iter() call it was doing into_iter() implicitly, which was taking ownership of the keys, vs using iter() just performs an immutable borrow instead.

Anyway, fixed in my working copy, much cleaner looking now.